src/mpi/coll/bcast/bcast_intra_scatter_recursive_doubling_allgather.c

```diff
@@ -224,7 +224,7 @@ int MPIR_Bcast_intra_scatter_recursive_doubling_allgather(void *buffer,
                     /* printf("Rank %d waiting to recv from rank %d\n",
                      * relative_rank, dst); */
                     mpi_errno = MPIC_Recv(((char *) tmp_buf + offset),
-                                          nbytes - offset,
+                                          nbytes - offset < 0 ? 0 : nbytes - offset,
```

Comment:
As I explained in https://github.com/pmodels/mpich/pull/5454#issuecomment-889485999, I don't think this is the right fix.

Reply:
I will make the change as you mentioned in the comment to not invoke Send or Recv when count becomes 0 or negative.

Reply:
Pushed the code that checks and ignores Send and Recv when count is 0 or negative. This introduces many if else branch checks. Since this is a corner case, that might not even occur, should we just `assert ` the count is positive and proceed without any condition check during Send and Recv. That would keep the code simpler.

Reply:
@srirajpaul You misunderstood. I believe the `nbytes - offset` is non-negative by algorithm, thus the check is unnecessary. The checks in the other places are also unnecessary and need to be removed.

My question was -- try find an example, `comm_size` and `bytes` combination that would result in negative `nbytes - offset`. You can do that by simply run the code. Then you can convince yourself that the check is unnecessary. Ideally, one should analyze the algorithm and prove the case.

Add unnecessary code misdirects the reader, and obscures the algorithm.

Reply:
@hzhou I also dont think the values can go negative now. But if I work back, it look like these checks are added more like a defensive programming. Since this is large codebase, incorrect changes in other modules in future may result in negative values here. I will go forward and close the PR.

Reply:
> Since this is large codebase, incorrect changes in other modules in future may result in negative values here.

It won't 😄 Try find a counter example if you don't believe it.

Reply:
May be like, someone incorrectly/accidently sets comm_ptr->local_size with a higher value, can lead to a higher relative_rank,  then higher relative_dst, then higher dst_tree_root, then higher recv_offset. Now can nbytes-recv_offset go -negative, I think.

Reply:
Inconsistent `comm` structure will cause more inconsistencies than you can catch here.

Reply:
Yes. I agree.
I was just telling that incorrect changes outside can lead to negative values and maybe that was the intention for the extra check. But as you mention, those can lead to more problems, which will be taken care by other test cases and so there is no need for the extra check here.

Reply:
@hzhou Here is a concrete example that gives -ve values. @tarudoodi helped to find the issue related to this PR. 
We start with with 6 ranks and the execution trace for rank 4.
Let count=1 and type_size=4 gives nbytes=4 and scatter_size=1 .
rank=4, relative_rank=4, gives relative_dest=5 when mask=1. Then i=0, root=0, comm_size=6, gives dst=5 and dst_tree_root=5.
with scatter_size=1 we get recv_offset=5 and now nbytes-recv_offset=-1

Reply:
I see. It happens when `nbytes < comm_size`. Thanks for digging up the context. 

Could you add a comment explaining the case? Probably you can just add the comment at line 103. https://github.com/pmodels/mpich/blob/814cd124573cee7bb02db3df280cbfbd63445217/src/mpi/coll/bcast/bcast_intra_scatter_recursive_doubling_allgather.c#L102-L104

Reply:
@hzhou  We came across this when we removed the power of 2 checks. After removing those checks and adding this patch for negative values, we were able to run this algorithm with non power of 2 ranks. It has been in our branch like that for over a year and we didn't notice any other bugs. Maybe we could remove those checks from `main` too?

Reply:
Yes, we can remove the power of 2 checks. There is no reason to have the code if we never test it.

Reply:
It seems some of my recent commits I made removed the options to reopen this PR. Let me create a new PR and refer to this one.